### PR TITLE
[FEAT] Product Search Endpoint - BACKEND #5

### PR DESCRIPTION
**Pull request by Am0du**
## Description
This feature involves creating a product search endpoint for the backend system. The endpoint will allow users to search for products by name, category, and price range. Additional search features can be added to enhance the functionality.
## Acceptance Criteria
- Users can search for products by name.
- Users can search for products by category.
- Users can filter search results by price range.
- The endpoint should return accurate search results based on the given criteria.
- The search functionality should be efficient and performant, even with a large dataset.
- Proper error handling and validation should be implemented for search queries.
## Task
- [x] Design the product search endpoint.
- [x] Implement query parameters for name, category, and price range.
- [x] Integrate the search endpoint with the existing product database.
- [x] Implement error handling and validation for search queries.
- [x] Write unit tests for the search functionality.
- [x] Write integration tests for the search endpoint.
- [x] Document the API endpoint and usage.
## Expected Outcome
- A functional product search endpoint that meets all acceptance criteria.
- Users can efficiently search for products using various criteria.
- The search results are accurate and relevant to the user's query.
- The system handles a high volume of search requests without significant performance degradation.

### CODE REVIEW

File: src/main/java/hng_java_boilerplate/product/controller/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            productSearchDTO.setProducts(ProductMapper.INSTANCE.toDTOList(products));
+            productSearchDTO.setSuccess(true);
+            return new ResponseEntity<>(productSearchDTO, HttpStatus.OK);
+        }
+
+        productSearchDTO.setStatus_code(HttpStatus.OK.value());
+        productSearchDTO.setProducts(ProductMapper.INSTANCE.toDTOList(products));
+        productSearchDTO.setSuccess(true);
+        return new ResponseEntity<>(productSearchDTO, HttpStatus.OK);
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/controller/ProductController.java:52]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/dto/ErrorDTO.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+public class ErrorDTO {
+    private String parameter;
+    private String message;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ErrorDTO.java:15]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/dto/ProductDTO.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ProductDTO {
+    private String id;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductDTO.java:13]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/dto/ProductErrorDTO.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ProductErrorDTO {
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductErrorDTO.java:13]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/dto/ProductSearchDTO.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ProductSearchDTO {
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductSearchDTO.java:14]: Remove extra spaces

File: src/main/java/hng_java_boilerplate/product/dto/ProductValidatorDTO.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package hng_java_boilerplate.product.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductValidatorDTO.java:9]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/entity/Product.java [hunk 7]
```diff
@@ -4,53 +4,29 @@
...
+    private double price;
+    @Column(name = "image_url")
+    private String imageUrl;
 
     @ManyToOne
     @JsonIgnore
     @JoinColumn(name = "user_id", nullable = false)
     private User user;
-
-    public String getId() {
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/entity/Product.java:25]: Why did you remove all the setters?

File: src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    public ResponseEntity<?> handleException(ValidationError v){
+        ProductValidatorDTO dto = new ProductValidatorDTO();
+        dto.setErrors(List.of(v.getError()));
+        dto.setStatus_code(HttpStatus.UNPROCESSABLE_ENTITY.value());
+        dto.setSuccess(false);
+
+        return new ResponseEntity<>(dto, HttpStatus.UNPROCESSABLE_ENTITY);
+
+    }
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/errorhandler/ProductErrorHandler.java:26]: Remove all unnecessary spaces

File: src/main/java/hng_java_boilerplate/product/exceptions/ValidationError.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+@Data
+public class ValidationError extends RuntimeException {
+
+    private ErrorDTO error;
+
+    public ValidationError(ErrorDTO error) {
+        this.error = error;
+    }
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/exceptions/ValidationError.java:16]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/product_mapper/ProductMapper.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+public interface ProductMapper {
+    ProductMapper INSTANCE = Mappers.getMapper(ProductMapper.class);
+
+    @Mapping(source = "id", target = "id")
+    @Mapping(source = "name", target = "name")
+    @Mapping(source = "category", target = "category")
+    @Mapping(source = "description", target = "description")
+    @Mapping(source = "imageUrl", target = "imageUrl")
+    ProductDTO toDTO(Product product);
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/product_mapper/ProductMapper.java:21]: Remove extra space

File: src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    @Override
+    public List<Product> productsSearch(String name, String category, Double minPrice, Double maxPrice) {
+        if (name == null || name.isEmpty()) {
+            ErrorDTO errorDTO = new ErrorDTO();
+            errorDTO.setMessage("name is a required parameter");
+            errorDTO.setParameter("name");
+            throw new ValidationError(errorDTO);
+        }
+
+            return productRepository.searchProducts(name, category, minPrice, maxPrice);
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java:31]: Remove extra spaces before the return statement

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

File: src/main/java/hng_java_boilerplate/product/entity/Product.java (hunk 7 above)
Comment (Am0du) [reply src/main/java/hng_java_boilerplate/product/entity/Product.java:25]: The @Data annotation provides the getters and setters

**Comment by NdubuisiJr** [comment]
This endpoint should be paginated. Add pagination parameters in the request. Check other issues to find the approved pagination models.

**Comment by Am0du** [comment]
> This endpoint should be paginated. Add pagination parameters in the request. Check other issues to find the approved pagination models.

i've added pagination

File: src/main/java/hng_java_boilerplate/product/dto/ProductErrorDTO.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+@Component
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class ProductErrorDTO {
+    private boolean success;
+    private String message;
+    private int status_code;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductErrorDTO.java:16]: Another space here

File: src/main/java/hng_java_boilerplate/product/dto/ProductSearchDTO.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+@AllArgsConstructor
+@Data
+public class ProductSearchDTO {
+    private boolean success;
+    private int status_code;
+    private List<ProductDTO> products;
+    private int total;
+    private int page;
+    private int limit;
+
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/dto/ProductSearchDTO.java:20]: Another space here

File: src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Override
+    public Page<Product> productsSearch(String name, String category, Double minPrice, Double maxPrice, Pageable pageable) {
+        if (name == null || name.isEmpty()) {
+            ErrorDTO errorDTO = new ErrorDTO();
+            errorDTO.setMessage("name is a required parameter");
+            errorDTO.setParameter("name");
+            throw new ValidationError(errorDTO);
+        }
+            return productRepository.searchProducts(name, category, minPrice, maxPrice, pageable);
```
Comment (NdubuisiJr) [inline src/main/java/hng_java_boilerplate/product/service/ProductServiceImpl.java:30]: Extra space before `return`

File: src/test/java/hng_java_boilerplate/product_test/e2e_test/ProductSearchTest.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
+//package hng_java_boilerplate.product_test.e2e_test;
```
Comment (NdubuisiJr) [inline src/test/java/hng_java_boilerplate/product_test/e2e_test/ProductSearchTest.java:1]: Delete this file entirely

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

**Comment by NdubuisiJr** [comment]
It's looking better. Resolve these few changes.

**Comment by Am0du** [comment]
> It's looking better. Resolve these few changes.

i've resolve all the changes.

**Review by NdubuisiJr — APPROVED** [review]
